File: data structures and algorithms/java/05_053.java
import java.util.*;
import java.util.stream.*;
import java.io.*;
import java.math.*;

public class Main {
    public static void main(String[] args) {
        // Main method code goes here
    }
    
    public class Node {
        public int key;
        public Node left;
        public Node right;
        
        public Node(int key) {
            this.key = key;
        }
    }
    
    public Node lowestCommonAncestor(Node root, Node p, Node q) {
        if (p.key < root.key && q.key < root.key) {
            return lowestCommonAncestor(root.left, p, q);
        } else if (p.key > root.key && q.key > root.key) {
            return lowestCommonAncestor(root.right, p, q);
        } else {
            return root;
        }
    }
}

File: data structures and algorithms/cpp/05_053.cpp
class Codec {
public:
    // Encodes a tree to a single string
    string serialize(Node* root) {
        if (!root) {
            return "null";
        }
        return std::to_string(root->value) + "," + 
               serialize(root->left) + "," + 
               serialize(root->right);
    }
    
    // Decodes your encoded data to tree
    Node* deserialize(string data) {
        std::istringstream stream(data);
        std::function<Node*()> dfs = [&]() -> Node* {
            std::string val;
            std::getline(stream, val, ',');
            
            if (val == "null") {
                return nullptr;
            }
            
            Node* node = new Node(std::stoi(val));
            node->left = dfs();
            node->right = dfs();
            return node;
        };
            
        return dfs();
    }
};